### PR TITLE
WIP: Full chacha parity with libsodium

### PR DESCRIPTION
**Pull request by emilbayes**
Still missing tests

### CODE REVIEW

File: binding.c [hunk 1]
```diff
@@ -699,6 +699,19 @@ napi_value sn_crypto_stream_xor(napi_env env, napi_callback_info info) {
   SN_RETURN(crypto_stream_xor(c_data, m_data, m_size, n_data, k_data), "stream encryption failed")
 }
 
+napi_value sn_crypto_stream_chacha20(napi_env env, napi_callback_info info) {
+  SN_ARGV(3, crypto_stream)
```
Comment (chm-diederichs) [inline binding.c]:
```suggestion
  SN_ARGV(3, crypto_stream_chacha20)
```

File: binding.c [hunk 2]
```diff
@@ -714,6 +727,66 @@ napi_value sn_crypto_stream_chacha20_xor (napi_env env, napi_callback_info info)
   SN_RETURN(crypto_stream_chacha20_xor(c_data, m_data, m_size, n_data, k_data), "stream encryption failed")
 }
 
+napi_value sn_crypto_stream_chacha20_xor_ic(napi_env env, napi_callback_info info) {
+  SN_ARGV(4, crypto_stream_xor)
```
Comment (chm-diederichs) [inline binding.c]:
```suggestion
  SN_ARGV(4, crypto_stream_chacha20_xor_ic)
```

File: binding.c [hunk 3]
```diff
@@ -2252,6 +2325,97 @@ napi_value sn_crypto_stream_chacha20_xor_wrap_final (napi_env env, napi_callback
...
+  state->block_counter = 0;
+  memcpy(state->n, n_data, crypto_stream_chacha20_ietf_NONCEBYTES);
+  memcpy(state->k, k_data, crypto_stream_chacha20_ietf_KEYBYTES);
+
+  return NULL;
+}
+
+
+napi_value sn_crypto_stream_chacha20_ietf_xor_wrap_update (napi_env env, napi_callback_info info) {
+  SN_ARGV(3, crypto_stream_chacha20_ietf_xor_instance_init)
```
Comment (chm-diederichs) [inline binding.c]:
```suggestion
  SN_ARGV(3, crypto_stream_chacha20_ietf_xor_wrap_update)
```

File: binding.c [hunk 4]
```diff
@@ -2252,6 +2325,97 @@ napi_value sn_crypto_stream_chacha20_xor_wrap_final (napi_env env, napi_callback
...
+typedef struct sn_crypto_stream_chacha20_ietf_xor_state {
+  unsigned char n[crypto_stream_chacha20_ietf_NONCEBYTES];
+  unsigned char k[crypto_stream_chacha20_ietf_KEYBYTES];
+  unsigned char next_block[64];
+  int remainder;
+  uint64_t block_counter;
+} sn_crypto_stream_chacha20_ietf_xor_state;
+
+napi_value sn_crypto_stream_chacha20_ietf_xor_wrap_init (napi_env env, napi_callback_info info) {
+  SN_ARGV(3, crypto_stream_chacha20_ietf_xor_instance_init)
```
Comment (chm-diederichs) [inline binding.c]:
```suggestion
  SN_ARGV(3, crypto_stream_chacha20_ietf_xor_wrap_init)
```

File: binding.c [hunk 5]
```diff
@@ -2252,6 +2325,97 @@ napi_value sn_crypto_stream_chacha20_xor_wrap_final (napi_env env, napi_callback
...
+    memcpy(c + m_size, next_block, state->remainder);
+
+    state->block_counter++;
+  }
+
+  return NULL;
+}
+
+napi_value sn_crypto_stream_chacha20_ietf_xor_wrap_final (napi_env env, napi_callback_info info) {
+  SN_ARGV(1, crypto_stream_chacha20_ietf_xor_instance_init)
```
Comment (chm-diederichs) [inline binding.c]:
```suggestion
  SN_ARGV(1, crypto_stream_chacha20_ietf_xor_wrap_final)
```

File: binding.c [hunk 6]
```diff
@@ -714,6 +727,66 @@ napi_value sn_crypto_stream_chacha20_xor (napi_env env, napi_callback_info info)
...
+
+  SN_THROWS(c_size != m_size, "m must be 'c.byteLength' bytes")
+  SN_ASSERT_LENGTH(n_size, crypto_stream_chacha20_NONCEBYTES, "n")
+  SN_ASSERT_LENGTH(k_size, crypto_stream_chacha20_KEYBYTES, "k")
+
+  SN_RETURN(crypto_stream_chacha20_xor_ic(c_data, m_data, m_size, n_data, ic, k_data), "stream encryption failed")
+}
+
+napi_value sn_crypto_stream_chacha20_ietf(napi_env env, napi_callback_info info) {
+  SN_ARGV(3, crypto_stream)
```
Comment (chm-diederichs) [inline binding.c]:
```suggestion
  SN_ARGV(3, crypto_stream_chacha20_ietf)
```

File: binding.c [hunk 7]
```diff
@@ -714,6 +727,66 @@ napi_value sn_crypto_stream_chacha20_xor (napi_env env, napi_callback_info info)
...
+  SN_ARGV_TYPEDARRAY(k, 2)
+
+  SN_ASSERT_LENGTH(n_size, crypto_stream_chacha20_ietf_NONCEBYTES, "n")
+  SN_ASSERT_LENGTH(k_size, crypto_stream_chacha20_ietf_KEYBYTES, "k")
+
+  SN_RETURN(crypto_stream_chacha20_ietf(c_data, c_size, n_data, k_data), "stream encryption failed")
+}
+
+napi_value sn_crypto_stream_chacha20_ietf_xor(napi_env env, napi_callback_info info) {
+  SN_ARGV(4, crypto_stream_xor)
```
Comment (chm-diederichs) [inline binding.c]:
```suggestion
  SN_ARGV(4, crypto_stream_chacha20_ietf_xor)
```

File: binding.c [hunk 8]
```diff
@@ -714,6 +727,66 @@ napi_value sn_crypto_stream_chacha20_xor (napi_env env, napi_callback_info info)
...
+
+  SN_THROWS(c_size != m_size, "m must be 'c.byteLength' bytes")
+  SN_ASSERT_LENGTH(n_size, crypto_stream_chacha20_ietf_NONCEBYTES, "n")
+  SN_ASSERT_LENGTH(k_size, crypto_stream_chacha20_ietf_KEYBYTES, "k")
+
+  SN_RETURN(crypto_stream_chacha20_ietf_xor(c_data, m_data, m_size, n_data, k_data), "stream encryption failed")
+}
+
+napi_value sn_crypto_stream_chacha20_ietf_xor_ic(napi_env env, napi_callback_info info) {
+  SN_ARGV(4, crypto_stream_xor)
```
Comment (chm-diederichs) [inline binding.c]:
```suggestion
  SN_ARGV(4, crypto_stream_chacha20_ietf_xor)
```

**Comment by mafintosh** [comment]
@emilbayes lets merge this first, then the experimental flag should fix the malloc warnings

**Pull request merged**